tree/treeplayer/inc/ROOT/TDFActionHelpers.hxx

```diff
@@ -303,12 +310,14 @@ public:
 
    void InitSlot(TTreeReader *, unsigned int) {}
 
-   void Exec(unsigned int slot, T v) { fColls[slot]->emplace_back(v); }
+   void Exec(unsigned int slot, T &v) { fColls[slot]->emplace_back(v); }
 
+   // This is optimised to treat vectors
    void Finalize()
    {
       ULong64_t totSize = 0;
-      for (auto &coll : fColls) totSize += coll->size();
+      for (auto &coll : fColls)
+         totSize += coll->size();
       auto rColl = fColls[0];
```

Comment:
I would expect `auto &rColl` here, how does this update the actual final result?

Reply:
Good comment. We could use a reference to optimise a bit, but fColls is a vector<shared_ptr<COLL>>, i.e. the way in which we access different slots.

Reply:
ah I forgot that we have `vec<shared_ptr<vec<T>>>` instead of `vec<vec<T>>`.
this is done to avoid a last copy from `fColls[0]` to the actual result object: we already make `fColls[0]` the actual result object.

I think with little effort we could actually have `vec<vec<T>>` and then do a `std::swap` between the result vector and `fColls[0]`, i'll write it down as a note for the future

Reply:
All performance optimisations count. In this case, I think we do not have extra copies but I may be mistaken. With N slots, the slots from 1 to N-1 are folded into slot 0 and that one is the one passed by reference in the constructor of the helper.

Reply:
I agree we are not making more copies than necessary, _maybe_ `vec<vec<T>>` instead of `vec<shared_ptr<vec<T>>>` could help with less impactful things (inlining, or when a CPU loads `TakeHelper` it already loads the vectors in the same cache line and doesn't have to go through a shared_ptr, things like that).

Reply:
fair enough.